lib/spack/spack/container/writers/__init__.py

```diff
@@ -0,0 +1,153 @@
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+"""Writers for different kind of recipes and related
+convenience functions.
+"""
+import collections
+import copy
+
+import spack.schema.env
+import spack.tengine as tengine
+import spack.util.spack_yaml as syaml
+
+from spack.container.images import build_info, package_info
+
+#: Caches all the writers that are currently supported
+_writer_factory = {}
+
+
+def writer(name):
```

Comment:
This feels like it's too clever and is going to become a maintenance problem (in terms of new people trying to read the code). Is there a performance concern about creating the writers at `.create()` time that motivates this factory solution?

Reply:
There are no performance concerns in this part of the code. The decorator is just a personal preference to be idiomatic when coding a writer:

https://github.com/spack/spack/blob/12fd35f2e6df6a9a9b1098f8a2ffdb63b8bdec33/lib/spack/spack/container/writers/docker.py#L10-L12

since it makes clear that something is a writer in the context where it's coded.

I used the term `factory` in the docstring since nothing prevents to use a function instead of a class if it has the right signature. For instance if we have a use case for something that manipulates the configuration before instantiating a docker writer right now we just need to add something like:
```python
@writer('docker-custom')
def docker_custom(configuration):
    # Modify configuration here
    return DockerContext(configuration)
```
in `docker.py` - no need to aggregate knowledge in `__init__.py`. I'm curious why do you think this is motivated by performance concerns?

Reply:
> Is there a performance concern about creating the writers at .create() time that motivates this factory solution?

I feel like I'm losing something here. Writers are currently created at `.create` time. The decorator is used to build a dictionary that maps the name of the "format" in the `spack.yaml` to the correct writer.